### PR TITLE
Refactor input resource volume handling to remove a type switch state…

### PR DESCRIPTION
**Pull request by dlorenc**
…ment.

# Changes

This change refactors the way input resources attach volumes to their steps.
Previously, there was a special-cased switch statement for Storage type resources.
This made the logic for resources hard to follow, and spread across the codebase. This
change refactors that by moving it up into the general Resource interface.

This is another follow-on to https://github.com/tektoncd/pipeline/pull/1135

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

### CODE REVIEW

File: pkg/apis/pipeline/v1alpha1/build_gcs_resource.go [hunk 1]
```diff
@@ -177,3 +177,7 @@ func getArtifactType(val string) (GCSArtifactType, error) {
 func (s *BuildGCSResource) GetUploadVolumeSpec(spec *TaskSpec) ([]corev1.Volume, error) {
 	return getStorageUploadVolumeSpec(s, spec)
 }
+
+func (s *BuildGCSResource) GetDownloadVolumeSpec(spec *TaskSpec) ([]corev1.Volume, error) {
+	return getStorageUploadVolumeSpec(s, spec)
```
Comment (sbwsg) [inline pkg/apis/pipeline/v1alpha1/build_gcs_resource.go]: v minor but would renaming to `getStorageTransferVolumeSpec` or maybe even just `getStorageVolumeSpec` or similar make sense given this is now also being used for the download volume?

File: pkg/reconciler/v1alpha1/taskrun/resources/input_resources.go [hunk 2]
```diff
@@ -113,25 +112,13 @@ func AddInputResource(
 			taskSpec.Steps = append(copyStepsFromPrevTasks, taskSpec.Steps...)
 			taskSpec.Volumes = append(taskSpec.Volumes, as.GetSecretsVolumes()...)
 		} else {
-			switch resource.GetType() {
```
Comment (afrittoli) [inline pkg/reconciler/v1alpha1/taskrun/resources/input_resources.go]: So happy to see this `switch` go away :)
